services/notifications/service/notifications_service.go

```diff
@@ -0,0 +1,461 @@
+	query := database.QuerySelector{
+		"userid": database.QuerySelector{
+			"$in": userid_list.UserIDs,
+		},
+	}
+
+	var devices []models.Device
+	err = db.FindAll("devices", query, &devices)
+
+	if err != nil {
+		return err
+	}
+
+	for _, device := range devices {
+		err = UnsubscribeDeviceFromTopic(topic, device)
+
+		if err != nil {
+			return err
+		}
+	}
+
+	return db.Update("topics", selector, &modifier)
+}
+
+/*
+	Links the given device token with a user
+*/
+func RegisterDeviceToUser(user_id string, device_reg models.DeviceRegistration) error {
+	var device_arn string
+	var platform_arn string
+
+	// Map platform (android, ios etc) to its ARN
+	switch strings.ToLower(device_reg.Platform) {
+	case "android":
+		platform_arn = config.ANDROID_PLATFORM_ARN
+	case "ios":
+		platform_arn = config.IOS_PLATFORM_ARN
+	default:
+		return errors.New("Invalid platform")
+	}
+
+	if config.IS_PRODUCTION {
+		out, err := client.CreatePlatformEndpoint(&sns.CreatePlatformEndpointInput{CustomUserData: &user_id, Token: &device_reg.DeviceToken, PlatformApplicationArn: &platform_arn})
+
+		if err != nil {
+			return err
+		}
+
+		device_arn = *out.EndpointArn
+	}
+
+	subs := make(map[string]string)
+	device := models.Device{DeviceArn: device_arn, DeviceToken: device_reg.DeviceToken, Platform: device_reg.Platform, UserID: user_id, Subscriptions: subs}
+
+	err := db.Insert("devices", device)
+
+	if err != nil {
+		return err
+	}
+
+	// Subscribe the device to all of a user's topics
+
+	topic_selector := database.QuerySelector{
+		"userids": database.QuerySelector{
+			"$all": [1]string{user_id},
+		},
+	}
+
+	var topics []models.Topic
+	err = db.FindAll("topics", topic_selector, &topics)
+
+	if err != nil {
+		return err
+	}
+
+	for _, topic := range topics {
+		err = SubscribeDeviceToTopic(topic, device)
+
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+/*
+	Subscribes a given Device to a Topic, in the database and SNS
+*/
+func SubscribeDeviceToTopic(topic models.Topic, device models.Device) error {
+	app_protocol := APPLICATION_PROTOCOL
+
+	var sub_arn string
+
+	if config.IS_PRODUCTION {
+		out, err := client.Subscribe(&sns.SubscribeInput{Protocol: &app_protocol, TopicArn: &topic.Arn, Endpoint: &device.DeviceArn})
+
+		if err != nil {
+			return err
+		}
+
+		sub_arn = *out.SubscriptionArn
+	}
+
+	device_selector := database.QuerySelector{
+		"devicearn": device.DeviceArn,
+	}
+
+	set_query := fmt.Sprintf("subscriptions.%s", topic.Name)
+
+	// Keep track of subscription's ARN so we can unsubscribe later
+	device_modifier := database.QuerySelector{
+		"$set": database.QuerySelector{
+			set_query: sub_arn,
+		},
+	}
+
+	err := db.Update("devices", device_selector, &device_modifier)
+
+	return err
+}
+
+/*
+	Unsubscribes a given Device from a Topic, both in the database and SNS
+*/
+func UnsubscribeDeviceFromTopic(topic models.Topic, device models.Device) error {
+	sub_arn, ok := device.Subscriptions[topic.Name]
+	if ok {
```

Comment:
Instead of wrapping this in an `if ok`, it is cleaner to do something like:
```
if !ok {
    return errors.New("Invalid topic name")
}
```
And then continue with the rest of the code.